File: problems/Copy List with Random Pointer/copyRandomList.cs
/*
// Definition for a Node.
public class Node {
    public int val;
    public Node next;
    public Node random;

    public Node(int _val) {
        val = _val;
        next = null;
        random = null;
    }
}
*/
public class Solution {
    // HashMap which holds old nodes as keys and new nodes as its values.
    private Dictionary<Node, Node> visitedHash = new Dictionary<Node, Node>();

    public Node CopyRandomList(Node head) {
        if (head == null) {
            return null;
        }

        // If we have already processed the current node, then we simply return the cloned version of
        // it.
        if (visitedHash.ContainsKey(head)) {
          return visitedHash[head];
        }

        // Create a new node with the value same as old node. (i.e. copy the node)
        Node node = new Node(head.val, null, null);

        // Save this value in the hash map. This is needed since there might be
        // loops during traversal due to randomness of random pointers and this would help us avoid
        // them.
        visitedHash.Add(head, node);

        // Recursively copy the remaining linked list starting once from the next pointer and then from
        // the random pointer.
        // Thus we have two independent recursive calls.
        // Finally we update the next and random pointers for the new node created.
        node.next = CopyRandomList(head.next);
        node.random = CopyRandomList(head.random);

        return node;
    }
}

File: problems/Copy List with Random Pointer/copyRandomList.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        // return copyRandomListRecursive(head);
        return copyRandomListIterative(head);
    }

    public Node copyRandomListIterative(Node head) {
        if (null == head) {
            return null;
        }

        HashMap<Node, Node> oldNodeToNewNodeMap = new HashMap<>();
        Node node = new Node(head.val);
        Node fakeHead = new Node(-1, node, null);

        oldNodeToNewNodeMap.put(head, node);

        while (null != head) {
            if (null != head.random) {
                if (!oldNodeToNewNodeMap.containsKey(head.random)) {
                    oldNodeToNewNodeMap.put(head.random, new Node(head.random.val, null, null));
                }

                node.random = oldNodeToNewNodeMap.get(head.random);
            }

            if (null != head.next) {
                if (!oldNodeToNewNodeMap.containsKey(head.next)) {
                    oldNodeToNewNodeMap.put(head.next, new Node(head.next.val, null, null));
                }

                node.next = oldNodeToNewNodeMap.get(head.next);
            }

            node = node.next;
            head = head.next;
        }

        return fakeHead.next;
    }

    public Node copyRandomListRecursive(Node head) {
        if (null == head) {
            return null;
        }

        Node newHead = new Node(head.val);
        HashMap<Node, Node> oldNodeToNewNodeMap = new HashMap<>();

        oldNodeToNewNodeMap.put(head, newHead);
        newHead.next = dfs(head.next, oldNodeToNewNodeMap);
        newHead.random = dfs(head.random, oldNodeToNewNodeMap);

        return newHead;
    }

    Node dfs(Node node, HashMap<Node, Node> oldNodeToNewNodeMap) {
        if (null == node) {
            return null;
        }

        if (!oldNodeToNewNodeMap.containsKey(node)) {
            Node newNode = new Node(node.val);

            oldNodeToNewNodeMap.put(node, newNode);
            newNode.next = dfs(node.next, oldNodeToNewNodeMap);
            newNode.random = dfs(node.random, oldNodeToNewNodeMap);
        }

        return oldNodeToNewNodeMap.get(node);
    }
}
